firmware/application/ui_external_items_menu_loader.cpp

```diff
@@ -0,0 +1,137 @@
+#include "ui_external_items_menu_loader.hpp"
+
+#include "sd_card.hpp"
+
+namespace ui {
+
+std::vector<Bitmap*> ExternalItemsMenuLoader::bitmaps;
+
+std::vector<GridItem> ExternalItemsMenuLoader::load_external_items(app_location_t app_location, NavigationView& nav) {
+    for (const auto& bitmap : bitmaps) {
+        delete bitmap->data;
+        delete bitmap;
+    }
+    bitmaps.clear();
+
+    std::vector<GridItem> exteral_apps;
+
+    if (sd_card::status() != sd_card::Status::Mounted)
+        return exteral_apps;
+
+    for (const auto& entry : std::filesystem::directory_iterator(u"APPS", u"*.ppma")) {
+        FIL ppma_file;
+        auto filePath = u"/APPS/" + entry.path().native();
+        if (f_open(&ppma_file, reinterpret_cast<const TCHAR*>(filePath.c_str()), FA_READ) != FR_OK)
+            continue;
+
+        size_t bytes_read;
+        application_information_t application_information = {};
+        if (f_read(&ppma_file, &application_information, sizeof(application_information_t), &bytes_read) != FR_OK) continue;
+
+        if (application_information.menu_location != app_location)
+            continue;
+
+        bool versionMatches = VERSION_MD5 == application_information.app_version;
+
+        GridItem gridItem = {};
+        gridItem.text = reinterpret_cast<char*>(&application_information.app_name[0]);
+
+        if (versionMatches) {
+            gridItem.color = Color((uint16_t)application_information.icon_color);
+
+            auto bitmapData = new uint8_t[32];
+            Bitmap* bitmap = new Bitmap{
```

Comment:
Probably better to wrap this up in a new type to make handling the allocated memory cleaner.
e.g.
```
struct DynamicBitmap : Bitmap {
  std::unique_ptr<std::array<uint8_t, 32>> data;
  DynamicBitmap(Size s, std::unique_ptr<std::array<uint8_t, 32>> d)
    Bitmap{s, d.get()}, data{std::move(d)} { }
};
```

Or something like that.

Reply:
Actually, it'd be better if that type just handled all of the buffer allocation directly.

Reply:
```cpp
template <size_t Width, size_t Height>
class DynamicBitmap {
   public:
    static constexpr uint8_t buffer_size = (Width / sizeof(uint8_t)) * Height;
    DynamicBitmap(const uint8_t* data)
        : _buffer(buffer_size, 0),
          _bitmap{new Bitmap{{Width, Height}, &_buffer[0]}}
    {
        memcpy(&_buffer[0], data, buffer_size);
    }

    const Bitmap* bitmap() { return _bitmap.get(); }

   private:
    // Allocating both members so the class is movable without invalidation.
    std::vector<uint8_t> _buffer;
    std::unique_ptr<Bitmap> _bitmap;
};
```

Reply:
Then used like this
```cpp
    uint8_t* ptr = nullptr;  // your ptr to raw bitmap data
    std::vector<DynamicBitmap<16, 16>> bitmaps;

    auto dyn_bmp = DynamicBitmap<16, 16>{ptr};
    bitmaps.push_back(std::move(dyn_bmp));

    const Bitmap* bitmap = dyn_bmp.bitmap();
```